### PR TITLE
Improve performance in the WP_Theme_JSON class by simplifying calls to array functions

### PR DESCRIPTION
**Pull request by aristath**
Trac ticket: https://core.trac.wordpress.org/ticket/56974

This PR simplifies expensive array-functions calls where appropriate.

Attaching some profiling results to showcase the impact.

Before (trunk):
Screenshot 2022-11-03 at 9 43 25 AM

After (with this PR):
Screenshot 2022-11-03 at 9 43 15 AM

All numbers in the screenshots above use milliseconds, and the impact of these seemingly simple changes is bigger than expected, shaving off more than 200ms. Of course those numbers are a bit inflated due to the fact that profiling slows down a page significantly, but the overall improvement on the page-load is more than 3%.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

props @jrfnl 🙌

### CODE REVIEW

File: src/wp-includes/class-wp-theme-json.php [hunk 1]
```diff
@@ -456,7 +456,7 @@ class WP_Theme_JSON {
 	public static function get_element_class_name( $element ) {
 		$class_name = '';
 
-		if ( array_key_exists( $element, static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES ) ) {
+		if ( isset( static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES[ $element ] ) ) {
```
Comment (jrfnl) [inline src/wp-includes/class-wp-theme-json.php]: > Fatal error: Cannot use isset() on the result of an expression (you can use "null !== expression" instead) in ./src/wp-includes/class-wp-theme-json.php on line 459

Comment (jrfnl) [reply src/wp-includes/class-wp-theme-json.php]: ☝🏻 This is your PHP 5.6 error ;-)

Comment (spacedmonkey) [reply src/wp-includes/class-wp-theme-json.php]: https://github.com/WordPress/gutenberg/pull/42567

Comment (jrfnl) [reply src/wp-includes/class-wp-theme-json.php]: Just as a note for the record: running `isset()`on plain constants/class constants is fine, it's only array access on those constants which is problematic.

The ability to declare a `const` as an array was only added in PHP 5.6, the ability to do the same using `define()` was added in PHP 7.0, which is probably why the ability to use `isset()` on array access on constants was added in PHP 7.0.

Comment (aristath) [reply src/wp-includes/class-wp-theme-json.php]: Perfect, thank you! I reverted the lines with the 5.6 issue, and added notes so we know what to do in the future after WP drops support for PHP 5.6 😄

**Comment by spacedmonkey** [comment]
While doing some profiling, I noticed there are still calls to `_wp_array_get`. This calls `array_key_exists`, which is expensive. I wonder if we could avoid these somehow.

File: src/wp-includes/class-wp-theme-json.php [hunk 2]
```diff
@@ -2317,6 +2343,9 @@ public function merge( $incoming ) {
 					) {
 						_wp_array_set( $this->theme_json, $path, $content );
 					} else {
+						$slugs_node = static::get_default_slugs( $this->theme_json, $node['path'] );
+						$slugs      = array_merge_recursive( $slugs_global, $slugs_node );
```
Comment (spacedmonkey) [inline src/wp-includes/class-wp-theme-json.php:2359]: Why is this needed?

**Comment by jrfnl** [comment]
> While doing some profiling, I noticed there are still calls to `_wp_array_get`. This calls `array_key_exists`, which is expensive. I wonder if we could avoid these somehow.

I wouldn't worry too much about (or overestimate) the impact of `array_key_exists()`.
* `array_key_exists()` is **_not_** one of the typically expensive `array_*()` function calls.
* Yes, `isset()` is generally speaking faster...
* ... BUT `array_key_exists()` has been [opcode optimized in PHP since PHP 7.4](https://speakerdeck.com/jrf/for-the-love-of-code-modernizing-wordpress-plugins-and-themes-full-deck?slide=45) and what with the [fast majority of WP users being on PHP 7.4 (or above)](https://wordpress.org/about/stats/), the performance difference will be negligible for most users.

Note: `isset()` and `array_key_exists()` also have a _functional_ difference in how they treat a `null` value, so for `array_key_exists()` to be equivalent to `isset()`, it MUST be accompanied by a `&& $var !== null`.

**Comment by aristath** [comment]
> Note: `isset()` and `array_key_exists()` also have a _functional_ difference in how they treat a `null` value, so for `array_key_exists()` to be equivalent to `isset()`, it MUST be accompanied by a `&& $var !== null`.

Yeah, that's an important distinction!
In the comments I added for future reference to replace `array_key_exists()` with `isset()` (`// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.`), I checked the values in each case and only added that comment in cases where the value can not be `null`. 
In one case the value _can_ be `null`, and I added a comment in that one that `// We need to use array_key_exists() instead of isset() because the value can be null.` 👍

File: src/wp-includes/class-wp-theme-json.php (hunk 2 above)
Comment (aristath) [reply src/wp-includes/class-wp-theme-json.php:2359]: It wasn't _added_, it was just _moved_ here. 😉 
A few lines above we were calling  these 2 on the previous `foreach` loop (used to be line 2281), so they were always running. Instead of always calling them, we moved them here where they are actually used. This way these rather expensive calls run fewer times, only when necessary.

**Comment by spacedmonkey** [comment]
In trunk. 
Screenshot 2022-11-09 at 21 54 06

In this brunch. 
Screenshot 2022-11-09 at 21 54 15

**Comment by felixarntz** [comment]
@jrfnl @aristath Following up on `array_key_exists`: Makes sense not to overthink this in terms of performance impact compared to `isset`, however I had one follow up thought I wanted to at least raise here:
* Most of the usage of `array_key_exists` here could be replaced with `isset` as the additional `null` consideration is irrelevant for most (if not all) of the code where it's used here.
* Given that, there would be the alternative option of assigning the result to a variable and then calling `isset` on that variable.

For example, instead of:
```php
if ( array_key_exists( $element, static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES ) ) {
}
```

Do:
```php
$element_class_names = static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES;
if ( isset( $element_class_names[ $element ] ) {
}
```

Again, not saying this is necessarily worth doing (and certainly not a blocker), but I'd be curious if the first or the last is generally faster. :)

**Review by felixarntz — APPROVED** [review]
The code changes here look solid to me!

I of course trust the benchmarking data provided in https://core.trac.wordpress.org/ticket/56974#comment:1, but it may additionally be helpful to quantify the performance gain here in milliseconds as well.

Given that parsing `theme.json` affects the `wp_head` action quite a bit (see also #3536), I thought I'd do a performance comparison there first. Based on 9 test runs, the median impact there is small but consistent:
* With Twenty Twenty-Three (i.e. block theme): 26.63ms vs 27.78ms (~4% faster)
* With Twenty Twenty-One (i.e. classic theme): 15.30ms vs 16.05ms (~5% faster)

See [this spreadsheet](https://docs.google.com/spreadsheets/d/1PSe9wx3xSnyQNS5w_NJUrHXo-txweOvt_RXCeYrvcLU/edit) for the full data.

So also in that regard it looks like a small but worthwhile win. Potentially I can do the same for more concretely the parsing of `theme.json` itself as that would be even closer to the changes here and allow us to quantify impact in that way too.

**Comment by jrfnl** [comment]
There's bound to be more room for improvement, but yes, let's take this step by step and I do believe this is a valuable step in that.

Re: assign+`isset` vs `array_key_exists`: as most `array_key_exists()` are only in place for PHP 5.6, might be more effective to re-open the discussion on dropping support for PHP 5.6 ... 😇 

On that note - PHP 8.3 will bring another very useful improvement [`json_validate()`](https://wiki.php.net/rfc/json_validate) (validate a JSON input before decoding without running the risk of the JSON tying up all the memory), but that is no doubt still a bridge too far ;-)

File: src/wp-includes/class-wp-theme-json.php [hunk 3]
```diff
@@ -603,6 +607,7 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
 		 */
 		$styles_non_top_level = static::VALID_STYLES;
 		foreach ( array_keys( $styles_non_top_level ) as $section ) {
+			// We need to use array_key_exists() instead of isset() because the value can be null.
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php]: Could this be reworded to remove "we"?

File: src/wp-includes/class-wp-theme-json.php [hunk 4]
```diff
@@ -456,6 +456,7 @@ class WP_Theme_JSON {
 	public static function get_element_class_name( $element ) {
 		$class_name = '';
 
+		// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php]: Could a Trac ticket be opened for tracking this with a detailed explanation of why that can be linked to for additional context? At a minimum, a summary of the discussion on this PR into a comment on the current Trac ticket would also work.

File: src/wp-includes/class-wp-theme-json.php [hunk 5]
```diff
@@ -626,6 +631,7 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
 		foreach ( $valid_element_names as $element ) {
 			$schema_styles_elements[ $element ] = $styles_non_top_level;
 
+			// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php]: Same here for using `See: URL.` to link off for additional details/context.

There are several more throughout the changes here, but I'm going to stop commenting on each one inline.

File: src/wp-includes/class-wp-theme-json.php [hunk 6]
```diff
@@ -1662,6 +1672,7 @@ protected static function compute_style_properties( $styles, $settings = array()
 			if ( is_array( $value_path ) ) {
 				$path_string = implode( '.', $value_path );
 				if (
+					// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php]: Same here for using `See: URL.` to link off for additional details/context.

**Review by desrosj — CHANGES_REQUESTED** [review]
Added a few nits.

File: src/wp-includes/class-wp-theme-json.php (hunk 3 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php]: Updated.

File: src/wp-includes/class-wp-theme-json.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php]: Opened https://core.trac.wordpress.org/ticket/57067 and updated all of these comments throughout the PR to point to it.

File: src/wp-includes/class-wp-theme-json.php [hunk 7]
```diff
@@ -1273,8 +1281,12 @@ protected function get_css_variables( $nodes, $origins ) {
 
 			$selector = $metadata['selector'];
 
-			$node         = _wp_array_get( $this->theme_json, $metadata['path'], array() );
-			$declarations = array_merge( static::compute_preset_vars( $node, $origins ), static::compute_theme_vars( $node ) );
+			$node                    = _wp_array_get( $this->theme_json, $metadata['path'], array() );
+			$declarations            = static::compute_preset_vars( $node, $origins );
+			$theme_vars_declarations = static::compute_theme_vars( $node );
+			foreach ( $theme_vars_declarations as $declaration ) {
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php]: I got a bit tripped up on these variable names. `$declaration` feels like `$declarations` is being iterated, but that's not the case here. Maybe `$theme_var_declaration` instead? Don't really feel strongly about this, it just read a bit weird to me.

File: src/wp-includes/class-wp-theme-json.php [hunk 8]
```diff
@@ -519,7 +521,10 @@ public function __construct( $theme_json = array(), $origin = 'theme' ) {
 		$nodes = static::get_setting_nodes( $this->theme_json );
 		foreach ( $nodes as $node ) {
 			foreach ( static::PRESETS_METADATA as $preset_metadata ) {
-				$path   = array_merge( $node['path'], $preset_metadata['path'] );
+				$path = $node['path'];
+				foreach ( $preset_metadata['path'] as $subpath ) {
+					$path[] = $subpath;
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php:526]: Just noting that this doesn't protect against duplicate values being added to `$path`, but that's how it works currently as `array_merge()` also results in duplicates values being present.

File: src/wp-includes/class-wp-theme-json.php [hunk 9]
```diff
@@ -2283,11 +2310,11 @@ public function merge( $incoming ) {
 		$nodes        = static::get_setting_nodes( $incoming_data );
 		$slugs_global = static::get_default_slugs( $this->theme_json, array( 'settings' ) );
 		foreach ( $nodes as $node ) {
-			$slugs_node = static::get_default_slugs( $this->theme_json, $node['path'] );
-			$slugs      = array_merge_recursive( $slugs_global, $slugs_node );
-
 			// Replace the spacing.units.
-			$path    = array_merge( $node['path'], array( 'spacing', 'units' ) );
+			$path   = $node['path'];
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php:2314]: Why not just use an array declaration here instead of three individual assignments?

```
$path = array(
	$node['path'],
	'spacing',
	'units'
);
```

File: src/wp-includes/class-wp-theme-json.php [hunk 10]
```diff
@@ -2298,19 +2325,25 @@ public function merge( $incoming ) {
...
 				foreach ( static::VALID_ORIGINS as $origin ) {
-					$base_path = array_merge( $node['path'], $preset['path'] );
-					$path      = array_merge( $base_path, array( $origin ) );
-					$content   = _wp_array_get( $incoming_data, $path, null );
+					$base_path = $node['path'];
+					foreach ( $preset['path'] as $leaf ) {
+						$base_path[] = $leaf;
+					}
+
+					$path   = $base_path;
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php:2333]: Same question here about why not use `$path = array()`;

File: src/wp-includes/class-wp-theme-json.php [hunk 11]
```diff
@@ -2463,7 +2504,8 @@ protected static function get_default_slugs( $data, $node_path ) {
 	 * @return string|null
 	 */
 	protected function get_name_from_defaults( $slug, $base_path ) {
-		$path            = array_merge( $base_path, array( 'default' ) );
+		$path            = $base_path;
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php:2507]: Same question about `$path = array();`

File: src/wp-includes/class-wp-theme-json.php [hunk 12]
```diff
@@ -2593,8 +2637,9 @@ protected static function remove_insecure_settings( $input ) {
 		$output = array();
 		foreach ( static::PRESETS_METADATA as $preset_metadata ) {
 			foreach ( static::VALID_ORIGINS as $origin ) {
-				$path_with_origin = array_merge( $preset_metadata['path'], array( $origin ) );
-				$presets          = _wp_array_get( $input, $path_with_origin, null );
+				$path_with_origin   = $preset_metadata['path'];
```
Comment (desrosj) [inline src/wp-includes/class-wp-theme-json.php:2640]: Can we use `$path_with_origin = array( value1, value2 )` here as well?

**Review by desrosj — APPROVED** [review]
Looks good overall. Commented with a few small questions/suggestions, but none of them are really to be considered blockers.

File: src/wp-includes/class-wp-theme-json.php (hunk 8 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php:526]: Right, was looking at this too, but indeed not a change in behavior, so this is fine as is.

File: src/wp-includes/class-wp-theme-json.php (hunk 7 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php]: SG, will update.

File: src/wp-includes/class-wp-theme-json.php (hunk 9 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php:2314]: Your suggested code would not be the same result. `$node['path']` is an array, so in your code the first element would be an array and the other two would be strings, so that would not be correct I think.

File: src/wp-includes/class-wp-theme-json.php (hunk 10 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php:2333]: Based on what the original code does, `$path` needs to include the items in `$base_path`, plus the single `$origin` item, so if we set `$path` to an empty array here, the behavior would change.

File: src/wp-includes/class-wp-theme-json.php (hunk 11 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php:2507]: See above, that would change the behavior.

File: src/wp-includes/class-wp-theme-json.php (hunk 9 above)
Comment (desrosj) [reply src/wp-includes/class-wp-theme-json.php:2314]: Ah, yes. You're right on this one. I think a few others it may result in the same. But not this one.

File: src/wp-includes/class-wp-theme-json.php (hunk 12 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-theme-json.php:2640]: Similar to https://github.com/WordPress/wordpress-develop/pull/3555#discussion_r1019751990, that would change the data, since `$preset_metadata['path']` is an array itself, so you would end up with an array in an array.

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/54804.

File: src/wp-includes/class-wp-theme-json.php [hunk 13]
```diff
@@ -2035,7 +2056,9 @@ public function get_styles_for_block( $block_metadata ) {
 					// the feature selector. This may occur when multiple block
 					// support features use the same custom selector.
 					if ( isset( $feature_declarations[ $feature_selector ] ) ) {
-						$feature_declarations[ $feature_selector ] = array_merge( $feature_declarations[ $feature_selector ], $new_feature_declarations );
+						foreach ( $new_feature_declarations as $new_feature_declaration ) {
+							$feature_declarations[ $feature_selector ][] = $feature_declaration;
```
Comment (oandregal) [inline src/wp-includes/class-wp-theme-json.php:2060]: @aristath @desrosj @felixarntz @spacedmonkey @jrfnl @hellofromtonya 

Shouldn't the variable assigned be `$new_feature_declaration` instead of `$feature_declaration`? This has been reported as problematic by the Gutenberg lint jobs ([see conversation](https://github.com/WordPress/gutenberg/pull/46579/files#r1049600844)).

I haven't looked at what regression this introduced but it'd be good to create a follow-up PR fixing it and adding a test case.

Comment (jrfnl) [reply src/wp-includes/class-wp-theme-json.php:2060]: @oandregal I concur. This is a bug which should be fixed.
```suggestion
							$feature_declarations[ $feature_selector ][] = $new_feature_declaration;
```

Comment (aristath) [reply src/wp-includes/class-wp-theme-json.php:2060]: 🤦 yeah, looks like this was a typo.. Good catch 👍

Comment (jrfnl) [reply src/wp-includes/class-wp-theme-json.php:2060]: Happens to the best of us ;-)

Comment (oandregal) [reply src/wp-includes/class-wp-theme-json.php:2060]: I'm not familiar with this code path, so asked folks to provide details for testing at https://github.com/WordPress/gutenberg/pull/46579#discussion_r1049600844 so we can add a new unit test and gauge how severe this is.

Comment (jrfnl) [reply src/wp-includes/class-wp-theme-json.php:2060]: AFAICS, this should go in the next 6.1 patch release.

Edit to clarify: I'm basing this assessment on the following:
* It's very clearly a bug with an obvious fix.
* It's likely a regression compared to previous behaviour.
